### ISSUE TITLE
Gremlin Console: Previously installed plugin won't show up in list

### ISSUE THREAD

**Issue by rutgervanwilligen** [tinkerpop/tinkerpop3#562]
(I suppose this is related to issue #482...?)

The Gremlin Console (both 3.0.0.M7 and 3.0.0-SNAPSHOT) does not show the Neo4j plugin in the list of installed plugins, and `:install com.tinkerpop neo4j-gremlin` needs to be run in each session. The resulting message is always `A module with the name neo4j-gremlin is already installed`; however, it results in the plugin showing up in the list and allows for its activation.

This behaviour occurs in both 3.0.0.M7 and 3.0.0-SNAPSHOT. Below is 3.0.0.M7's output.

```
$ bin/gremlin.sh

         \,,,/
         (o o)
-----oOOo-(3)-oOOo-----
plugin activated: tinkerpop.server
plugin activated: tinkerpop.sugar
plugin activated: tinkerpop.tinkergraph
gremlin> :plugin use tinkerpop.neo4j
==>tinkerpop.neo4j could not be found - use ':plugin list' to see available plugins
gremlin> :plugin list
==>tinkerpop.server[active]
==>tinkerpop.gephi
==>tinkerpop.utilities
==>tinkerpop.sugar[active]
==>tinkerpop.tinkergraph[active]
gremlin> :install com.tinkerpop neo4j-gremlin 3.0.0.M7
==>A module with the name neo4j-gremlin is already installed
gremlin> :plugin list
==>tinkerpop.server[active]
==>tinkerpop.gephi
==>tinkerpop.utilities
==>tinkerpop.sugar[active]
==>tinkerpop.tinkergraph[active]
==>tinkerpop.neo4j
gremlin> :plugin use tinkerpop.neo4j
==>tinkerpop.neo4j activated
gremlin> 
```

**Comment by spmallette** [comment]
I'm not sure that i see the problem here.  This seems to work as expected.   This is a freshly downloaded version of M7:

```text
$ bin/gremlin.sh

         \,,,/
         (o o)
-----oOOo-(3)-oOOo-----
plugin activated: tinkerpop.server
plugin activated: tinkerpop.utilities
plugin activated: tinkerpop.tinkergraph
gremlin> :plugin list
==>tinkerpop.server[active]
==>tinkerpop.gephi
==>tinkerpop.utilities[active]
==>tinkerpop.sugar
==>tinkerpop.tinkergraph[active]
gremlin> :plugin use tinkerpop.neo4j
==>tinkerpop.neo4j could not be found - use ':plugin list' to see available plugins
```

The above is expected - neo4j is not packaged with the gremlin console.  to continue with that console session:

```text
gremlin> :install com.tinkerpop neo4j-gremlin 3.0.0.M7
==>Loaded: [com.tinkerpop, neo4j-gremlin, 3.0.0.M7]
gremlin> :plugin list
==>tinkerpop.server[active]
==>tinkerpop.gephi
==>tinkerpop.utilities[active]
==>tinkerpop.sugar
==>tinkerpop.tinkergraph[active]
==>tinkerpop.neo4j
gremlin> :plugin use tinkerpop.neo4j
==>tinkerpop.neo4j activated
gremlin> :plugin list
==>tinkerpop.server[active]
==>tinkerpop.gephi
==>tinkerpop.utilities[active]
==>tinkerpop.sugar
==>tinkerpop.tinkergraph[active]
==>tinkerpop.neo4j[active]
gremlin> :x
```

So we've installed/activated the plugin.  I kill the console and restart:

```text
$ bin/gremlin.sh

         \,,,/
         (o o)
-----oOOo-(3)-oOOo-----
plugin activated: tinkerpop.server
plugin activated: tinkerpop.utilities
plugin activated: tinkerpop.neo4j
plugin activated: tinkerpop.tinkergraph
gremlin> :plugin list
==>tinkerpop.server[active]
==>tinkerpop.gephi
==>tinkerpop.utilities[active]
==>tinkerpop.sugar
==>tinkerpop.neo4j[active]
==>tinkerpop.tinkergraph[active]
```

Note that the Neo4j plugin is active - am i missing something?

**Comment by spmallette** [comment]
hmm - actually disregard the previous comment to some degree.  i think i see what you are saying now.  what i don't understand is how you got the console into that state?  can you send additional reproduction steps that include the start/stop of the console (starting from a fresh download, preferrably)?

**Comment by dkuppitz** [comment]
Maybe it's worth to mention that I just ran into the same(?) problem. If you build everything from source, don't start the shell from the project root directory, do this instead:

```shell
tinkerpop $ cd gremlin-console/target/gremlin-console-3.0.0-SNAPSHOT-standalone
tinkerpop/gremlin-console/target/gremlin-console-3.0.0-SNAPSHOT-standalone $ bin/gremlin.sh
```

**Comment by rutgervanwilligen** [comment]
Right, I think the problem has its root in the location from where the console is started. Up to now, I have always started the console by executing `$ bin/gremlin.sh` from within the `gremlin-console` folder. This is the produced output when installing the plugin at first (sorry, completely missed this one in my bug report):

```
$ bin/gremlin.sh

         \,,,/
         (o o)
-----oOOo-(3)-oOOo-----
gremlin> :plugin list
==>tinkerpop.server
==>tinkerpop.gephi
==>tinkerpop.utilities
==>tinkerpop.sugar
==>tinkerpop.tinkergraph
gremlin> :install com.tinkerpop neo4j-gremlin 3.0.0.M7
Detected a non-standard Gremlin directory structure during install.  Expecting a 'lib' directory sibling to 'ext'. This message does not necessarily imply failure, however the console requires a certain directory structure for proper execution. Altering that structure can lead to unexpected behavior.
==>Loaded: [com.tinkerpop, neo4j-gremlin, 3.0.0.M7]
gremlin> :plugin list
==>tinkerpop.server
==>tinkerpop.gephi
==>tinkerpop.utilities
==>tinkerpop.sugar
==>tinkerpop.tinkergraph
==>tinkerpop.neo4j
gremlin> :plugin use tinkerpop.neo4j
==>tinkerpop.neo4j activated
gremlin> :plugin list
==>tinkerpop.server
==>tinkerpop.gephi
==>tinkerpop.utilities
==>tinkerpop.sugar
==>tinkerpop.tinkergraph
==>tinkerpop.neo4j[active]
gremlin> 
```

Killing the console and restarting from the same location results in the behaviour I reported in my original post.

However, running `bin/gremlin.sh` from the `gremlin-console/target/gremlin-console-3.0.0.M7-standalone` folder produces your (i.e. the correct) behaviour.

So, apparently, when installing plugins, the directory which they're installed to is relative to the directory from which the console is started (either `tp3/bin/gremlin.sh`, `tp3/gremlin-console/bin/gremlin.sh` or `tp3/gremlin-console/target/gremlin-console-*standalone/bin/gremlin.sh`), leading to strange behaviour.

**Comment by spmallette** [comment]
I think this issue is resolved by setting `tinkerpop.ext` in `JAVA_OPTIONS`.  `tinkerpop.ext` was added a few weeks ago:

https://github.com/apache/incubator-tinkerpop/commit/ae148798e2852f4d9201399e68bc90439cbe6975

I just pushed this:

https://github.com/apache/incubator-tinkerpop/commit/780f3ce4a7a93057ff0b0ba9aedd5e65cec3362b

which seems to allow it to work from anywhere.  we are still working with apache to resolve our ability to deploy snapshots, so you might not be able to install the neo4j plugin from SNAPSHOT at this time if you wanted to try things out.
